src/charm.py

```diff
@@ -152,11 +153,59 @@ def __init__(self, *args) -> None:
     ###################
 
     def _on_livepatch_relation_changed(self, event) -> None:
-        """Handle peer `relation-changed` event."""
-        self._config_changed(event)
+        """
+        On peer relation changed hook.
+
+        This hook is for the non-leader units to get notified when the state
+        changes. On the leader unit this hook should be ignored to avoid
+        repetitive workload restarts while handling relations. This also means,
+        on the leader unit, whenever the state changes, the update workload
+        method should be called manually.
+        """
+        if self.unit.is_leader():
+            return
+        if not self._state.is_ready():
+            logging.warning("State is not ready")
+            return
+        self._update_workload_configuration(event)
+
+    def _restart_non_leader_units(self):
+        """
+        Cue non-leader units to update their workload configuration.
+
+        This method is meant to be used by action hooks that *do not* update the
+        state. It's because upon updating the state, the non-leader units will
+        get receive the cue via the peer-relation-changed event.
+        """
+        if not self.unit.is_leader() or not self._state.is_ready():
+            return
+        # By incrementing a value in the state, an peer-relation-changed event
+        # will be fired and the non-leader units will respond to it.
+        self._state.restart_cue = 1 + (self._state.restart_cue or 0)
+
+    def _restart_all_units(self):
+        """
+        Restart the leader along with non-leader units.
+
+        This method is meant to be used by action hooks that *do not* update the
+        state. It's because upon updating the state, the non-leader units will
+        get receive the cue via the peer-relation-changed event.
+        """
+        if not self.unit.is_leader() or not self._state.is_ready():
```

Comment:
if the unit is_leader, then this will return on L195  and NOT restart the unit..